src/SDKs/Azure.Base/data-plane/Azure.Base/Http/HeaderValues.cs

```diff
@@ -0,0 +1,517 @@
+    {
+        private static readonly string[] EmptyArray = new string[0];
+        public static readonly HeaderValues Empty = new HeaderValues(EmptyArray);
+
+        private readonly string _value;
+        private readonly string[] _values;
+
+        public HeaderValues(string value)
+        {
+            _value = value;
+            _values = null;
+        }
+
+        public HeaderValues(string[] values)
+        {
+            _value = null;
+            _values = values;
+        }
+
+        public static implicit operator HeaderValues(string value)
+        {
+            return new HeaderValues(value);
+        }
+
+        public static implicit operator HeaderValues(string[] values)
+        {
+            return new HeaderValues(values);
+        }
+
+        public static implicit operator string (HeaderValues values)
+        {
+            return values.GetStringValue();
+        }
+
+        public static implicit operator string[] (HeaderValues value)
+        {
+            return value.GetArrayValue();
+        }
+
+        public int Count => _value != null ? 1 : (_values?.Length ?? 0);
+
+        bool ICollection<string>.IsReadOnly => true;
+
+        string IList<string>.this[int index]
+        {
+            get => this[index];
+            set => throw new NotSupportedException();
+        }
+
+        public string this[int index]
+        {
+            get
+            {
+                if (_values != null)
+                {
+                    return _values[index]; // may throw
+                }
+                if (index == 0 && _value != null)
+                {
+                    return _value;
+                }
+                return EmptyArray[0]; // throws
+            }
+        }
+
+        public override string ToString()
+        {
+            return GetStringValue() ?? string.Empty;
+        }
+
+        private string GetStringValue()
+        {
+            if (_values == null)
+            {
+                return _value;
+            }
+            switch (_values.Length)
+            {
+                case 0: return null;
+                case 1: return _values[0];
+                default: return string.Join(",", _values);
+            }
+        }
+
+        public string[] ToArray()
+        {
+            return GetArrayValue() ?? EmptyArray;
+        }
+
+        private string[] GetArrayValue()
+        {
+            if (_value != null)
+            {
+                return new[] { _value };
+            }
+            return _values;
+        }
+
+        int IList<string>.IndexOf(string item)
+        {
+            return IndexOf(item);
+        }
+
+        private int IndexOf(string item)
+        {
+            if (_values != null)
+            {
+                var values = _values;
+                for (int i = 0; i < values.Length; i++)
+                {
+                    if (string.Equals(values[i], item, StringComparison.Ordinal))
```

Comment:
Given that HTTP headers themselves are case-insensitive, you may wish to consider treating them similarly here.    

_([RFC 7230, Section 3.2](https://tools.ietf.org/html/rfc7230#section-3.2))_

Reply:
It wouldn't be correct, only header names are defined as case-insensitive.

Reply:
It's true that only header names are called out by the spec to formally be case-insensitive, but, so far as I'm aware, the case-insensitivity for values is undefined.  In my experience, the case-insensitivity typically is extended to cover header/value pairings.

Reply:
It depends on the header. See for example https://www.w3.org/Protocols/rfc1341/4_Content-Type.html 
```
The type, subtype, and parameter names are not case sensitive. For example, TEXT, Text, and TeXt are all equivalent. Parameter values are normally case sensitive, but certain parameters are interpreted to be case- insensitive, depending on the intended use. (For example, multipart boundaries are case-sensitive, but the "access- type" for message/External-body is not case-sensitive.) 
```

Reply:
Fair enough for content negotiation.  Though, I'll mention that 1341 is the MIME protocol and isn't necessarily general.   That said, I have no strong opinions on the change, so I withdraw my objection.